### PR TITLE
bugfix: EnhancedServiceLoader may throw NoClassDefFoundError terminate class load

### PR DESCRIPTION
**Pull request by zhuyoufeng**
### Ⅰ. Describe what this PR did
When class loader loads the extensions from META-INF.services, System may throw NoClassDefFoundError and the EnhancedServiceLoader will stop to load the other classes.

### CODE REVIEW

File: common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java [hunk 1]
```diff
@@ -38,6 +24,15 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.net.URL;
+import java.util.*;
```
Comment (slievrly) [inline common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java]: forbid to import *

**Comment by slievrly** [comment]
why not fail fast？

**Comment by zhuyoufeng** [comment]
> why not fail fast

For example, a system use consul as registry server and doesn't dependent on other providers such as nacos , seata will load the the configuration provider from META-INF/services/io.seata.config.ConfigurationProvider line by line, and throw noclassdeffounderror then ternimate the extension loading because it cannot find some class such as nacos, but actually system doesn't need these classes.

image

File: common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java (hunk 1 above)
Comment (zhuyoufeng) [reply common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java]: updated, thanks

**Comment by slievrly** [comment]
> > why not fail fast
> 
> For example, a system use consul as registry server and doesn't dependent on other providers such as nacos , seata will load the the configuration provider from META-INF/services/io.seata.config.ConfigurationProvider line by line, and throw noclassdeffounderror then ternimate the extension loading because it cannot find some class such as nacos, but actually system doesn't need these classes.
> 
> image

EnhancedServiceLoader `class not found` means just like `io.seata.discovery.registry.nacos.NacosRegistryProvider` not found, but nacos-client not found. In which case, `io.seata.discovery.registry.nacos.NacosRegistryProvider` will not exist?

**Comment by zhuyoufeng** [comment]
> > > why not fail fast
> > 
> > 
> > For example, a system use consul as registry server and doesn't dependent on other providers such as nacos , seata will load the the configuration provider from META-INF/services/io.seata.config.ConfigurationProvider line by line, and throw noclassdeffounderror then ternimate the extension loading because it cannot find some class such as nacos, but actually system doesn't need these classes.
> > image
> 
> EnhancedServiceLoader `class not found` means just like `io.seata.discovery.registry.nacos.NacosRegistryProvider` not found, but nacos-client not found. In which case, `io.seata.discovery.registry.nacos.NacosRegistryProvider` will not exist?


not NacosRegistryProvider, this class is OK.

NacosConfigurationProvider, should be this, because this class depend on the NacosException, as the attached image, or no need to modify the EnhancedClassLoader, just modify the class NacosConfigurationProvider will be fine.

image

image

**Comment by slievrly** [comment]
@zhuyoufeng Thank you very much, I will reproduce this question.

**Comment by slievrly** [comment]
@zhuyoufeng The basic problem I think is that we should not have specific third-party dependencies in the `*Provider` class. Just like throwing Exception here

**Comment by zhuyoufeng** [comment]
@slievrly My pleasure ^_^, i closed this PR

**Pull request closed without merging**